File: LoochaMediaLibrary/src/com/realcloud/media/loader/VideoLoader.java
package com.realcloud.media.loader;

import android.content.Context;
import android.database.Cursor;
import android.provider.MediaStore;
import android.text.TextUtils;

import com.realcloud.media.MediaConstant;
import com.realcloud.media.model.Album;
import com.realcloud.media.model.Photo;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by zack on 2017/10/12.
 * 查询视频的loader
 */

public class VideoLoader extends PhotoLoader {

    private static final String[] PROJECTION_VIDEO = {MediaStore.Video.Media._ID, MediaStore.Video.Media.DISPLAY_NAME,MediaStore.Video.Media.DATA,MediaStore.Video.Media.DURATION
            ,MediaStore.Video.Media.WIDTH,MediaStore.Video.Media.HEIGHT};
    private static final String ORDER_BY = MediaStore.Video.Media.DATE_ADDED + " DESC";
    private static final String IS_LARGE_SIZE = MediaStore.Video.Media.SIZE+" > 0 ";

    public VideoLoader(Context context,OnPhotoLoaderListener onPhotoLoaderListener) {
        super(context,onPhotoLoaderListener);
    }

    @Override
    public List<Photo> loadInBackground() {
        List<Photo> photos = new ArrayList<>();
        String albumId = getBundleArgs().getString(MediaConstant.ALBUMID);
        if (!TextUtils.isEmpty(albumId)){
            if (albumId.equals(String.valueOf(Album.ALL_VIDEO)) || albumId.equals(String.valueOf(Album.ALL_PHOTO_VIDEO))){ //所有视频文件
                Cursor videoCusrsor = getContext().getApplicationContext().getContentResolver().query(MediaStore.Video.Media.EXTERNAL_CONTENT_URI
                        ,PROJECTION_VIDEO,IS_LARGE_SIZE ,null, ORDER_BY);
                if (videoCusrsor != null && videoCusrsor.getCount() > 0 && videoCusrsor.moveToFirst()){
                    do {
                        long id = videoCusrsor.getLong(videoCusrsor.getColumnIndex(MediaStore.Video.Media._ID));
                        String name = videoCusrsor.getString(videoCusrsor.getColumnIndex(MediaStore.Video.Media.DISPLAY_NAME));
                        String data = videoCusrsor.getString(videoCusrsor.getColumnIndex(MediaStore.Video.Media.DATA));
                        long duration = videoCusrsor.getLong(videoCusrsor.getColumnIndex(MediaStore.Video.Media.DURATION));
                        int width = videoCusrsor.getInt(videoCusrsor.getColumnIndex(MediaStore.Video.Media.WIDTH));
                        int height = videoCusrsor.getInt(videoCusrsor.getColumnIndex(MediaStore.Video.Media.HEIGHT));
                        Photo photo = new Photo(id,albumId,name,data,duration,width,height,getVideoThumbPath(id));
                        if (!photos.contains(photo)){
                            photos.add(photo);
                        }
                    }while (videoCusrsor.moveToNext());
                    videoCusrsor.close();
                }
            }else{
                Cursor videoCusrsor = getContext().getApplicationContext().getContentResolver().query(MediaStore.Video.Media.EXTERNAL_CONTENT_URI
                        ,PROJECTION_VIDEO,MediaStore.Video.Media.BUCKET_ID + " = ? and (" + IS_LARGE_SIZE + ")",new String[]{albumId}, ORDER_BY);
                if (videoCusrsor != null && videoCusrsor.getCount() > 0 && videoCusrsor.moveToFirst()){
                    do {
                        long id = videoCusrsor.getLong(videoCusrsor.getColumnIndex(MediaStore.Video.Media._ID));
                        String name = videoCusrsor.getString(videoCusrsor.getColumnIndex(MediaStore.Video.Media.DISPLAY_NAME));
                        String data = videoCusrsor.getString(videoCusrsor.getColumnIndex(MediaStore.Video.Media.DATA));
                        long duration = videoCusrsor.getLong(videoCusrsor.getColumnIndex(MediaStore.Video.Media.DURATION));
                        int width = videoCusrsor.getInt(videoCusrsor.getColumnIndex(MediaStore.Video.Media.WIDTH));
                        int height = videoCusrsor.getInt(videoCusrsor.getColumnIndex(MediaStore.Video.Media.HEIGHT));
                        Photo photo = new Photo(id,albumId,name,data,duration,width,height,getVideoThumbPath(id));
                        if (!photos.contains(photo)){
                            photos.add(photo);
                        }
                    }while (videoCusrsor.moveToNext());
                    videoCusrsor.close();
                }
            }
        }
        return photos;
    }
}

File: MediaLibrary/src/main/java/com/zack/media/loader/VideoLoader.kt
package com.zack.media.loader

import android.content.Context
import android.database.Cursor
import android.provider.MediaStore
import androidx.loader.content.Loader
import com.zack.media.model.AlbumBean
import com.zack.media.model.MediaBean
import com.zack.media.model.MediaType
import java.util.*

/**
 * @Author zack
 * @Date 2019/7/26
 * @Description 视频loader
 * @Version 1.0
 */
class VideoLoader(context: Context,val loaderCompleteCallback:(List<MediaBean>,List<AlbumBean>)->Unit) : MediaBaseLoader<Pair<List<MediaBean>,List<AlbumBean>>>(context) {

    override fun loadInBackground(): Pair<List<MediaBean>,List<AlbumBean>>? {
        val videoList = ArrayList<MediaBean>()
        val albumList = ArrayList<AlbumBean>()
        val albumId = getBundleArgs()?.getString(MediaLoaderConstant.ALBUM_ID)
        val videoCursor : Cursor?
        if (albumId.isNullOrEmpty()){
            videoCursor = context.applicationContext.contentResolver.query(
                MediaStore.Video.Media.EXTERNAL_CONTENT_URI,MediaLoaderConstant.VIDEO_PROJECTION,MediaLoaderConstant.VIDEO_SELECTION,
                null,MediaLoaderConstant.VIDEO_ORDER_BY)
        }else{
            val selection = String.format(MediaLoaderConstant.VIDEO_BUCKET_SELECTION,albumId)
            videoCursor = context.applicationContext.contentResolver.query(
                MediaStore.Video.Media.EXTERNAL_CONTENT_URI,MediaLoaderConstant.VIDEO_PROJECTION,selection,
                null,MediaLoaderConstant.VIDEO_ORDER_BY)
        }
        if (videoCursor != null && videoCursor.count > 0 && videoCursor.moveToFirst()){
            do {
                val id = videoCursor.getLong(videoCursor.getColumnIndex(MediaStore.Video.Media._ID))
                val bucketId = videoCursor.getLong(videoCursor.getColumnIndex(MediaStore.Video.Media.BUCKET_ID))
                val bucketName = videoCursor.getString(videoCursor.getColumnIndex(MediaStore.Video.Media.BUCKET_DISPLAY_NAME))
                val data = videoCursor.getString(videoCursor.getColumnIndex(MediaStore.Video.Media.DATA))
                val time = videoCursor.getLong(videoCursor.getColumnIndex(MediaStore.Video.Media.DATE_TAKEN))
                val duration = videoCursor.getLong(videoCursor.getColumnIndex(MediaStore.Video.Media.DURATION))
                val videoBean = MediaBean(id, bucketId, bucketName, data, time,duration = duration,mediaType = MediaType.TYPE_VIDEO)
                if (!videoList.contains(videoBean)) {
                    videoList.add(videoBean)
                    val albumBean = AlbumBean(videoBean.bucketId,videoBean.bucketName,videoBean.filePath,MediaType.TYPE_VIDEO)
                    if (!albumList.contains(albumBean)){
                        albumList.add(albumBean)
                    }
                }
            } while (videoCursor.moveToNext())
            videoCursor.close()
        }
        return Pair(videoList,albumList)
    }

    override fun onLoadFinished(loader: Loader<Pair<List<MediaBean>,List<AlbumBean>>>, data: Pair<List<MediaBean>,List<AlbumBean>>?) {
        val videoList = data?.first?:Collections.emptyList()
        val albumList = data?.second?:Collections.emptyList()
        loaderCompleteCallback(videoList,albumList)
    }
}